domain/controllerupgrader/state/controller.go

```diff
@@ -0,0 +1,169 @@
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	rval := make(map[string]semversion.Number, len(dbValues))
+	for _, v := range dbValues {
+		version, err := semversion.Parse(v.Version)
+		if err != nil {
+			return nil, errors.Errorf(
+				"parsing controller node %q agent version %q: %w",
+				v.ControllerID, v.Version, err,
+			)
+		}
+
+		rval[v.ControllerID] = version
+	}
+
+	return rval, nil
+}
+
+// GetControllerVersion returns the target controller version in use by the
+// cluster.
+func (s *ControllerState) GetControllerVersion(ctx context.Context) (semversion.Number, error) {
+	db, err := s.DB()
+	if err != nil {
+		return semversion.Number{}, errors.Capture(err)
+	}
+
+	var versionValue controllerTargetVersion
+	stmt, err := s.Prepare(`
+SELECT &controllerTargetVersion.*
+FROM   controller
+`,
+		versionValue)
+	if err != nil {
+		return semversion.Number{}, errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt).Get(&versionValue)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.New("no controller target version has been previously set")
+		}
+		return err
+	})
+
+	if err != nil {
+		return semversion.Zero, errors.Capture(err)
+	}
+
+	rval, err := semversion.Parse(versionValue.TargetVersion)
+	if err != nil {
+		return semversion.Zero, errors.Errorf(
+			"parsing target version %q for controller: %w",
+			versionValue.TargetVersion, err,
+		)
+	}
+
+	return rval, nil
+}
+
+// SetControllerTargetVersion is responsible for setting the current clusters
+// target controller version. Controllers in the cluster will eventually
+// upgrade to this version once changed.
+//
+// This func expects that a controller version has already been set. If this is
+// not the case no update will be performed and an error will be returned. It is
+// not the responsibility of this function to establish the initial controller
+// information.
+func (s *ControllerState) SetControllerTargetVersion(
+	ctx context.Context,
+	version semversion.Number,
+) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	toVersionInput := setControllerTargetVersion{
+		TargetVersion: version.String(),
+	}
+	stmt, err := s.Prepare(`
+UPDATE controller
+SET    target_version = $setControllerTargetVersion.target_version
+`,
+		toVersionInput)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	var outcome sqlair.Outcome
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, stmt, toVersionInput).Get(&outcome)
+	})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	updateCount, err := outcome.Result().RowsAffected()
+	if err != nil {
+		return errors.Errorf("getting update count after setting controller version: %w", err)
+	}
+	if updateCount == 0 {
+		return errors.New("no controller version has been previously set")
```

Comment:
suggestion: does this case happen because target version and current version of the controller are the same? I realised in the function `validateControllerCanBeUpgradedTo()` we do not check if the versions are the same as well. Perhaps if we could do the check further up in the validation logic, we could avoid the unnecessary state calls following as well as returning an error here? Or if the validation for same version had been done elsewhere we could avoid this err message? Because the error message "no controller version has been previously set" doesn't seem to explicitly state why the update was not set and im not sure if it'll be meaningful to return this message, what do you think?

Reply:
I see，thanks for the clarifications tom!